File: modules/beauty-filter/ios/BeautyFilterModule.swift
//
//  BeautyFilterModule.swift
//  BeautyFilter
//
//  Expo module definition. Contract is identical to the Android
//  BeautyFilterModule.kt so the shared JS (index.ts) works on both platforms.
//

import ExpoModulesCore

public class BeautyFilterModule: Module {
    public func definition() -> ModuleDefinition {
        Name("BeautyFilter")

        Function("isAvailable") { () -> Bool in
            true
        }

        // The live Metal pipeline already bakes the beauty filter into captures,
        // so post-processing is a no-op that returns the uri unchanged. This
        // exists purely for API parity with the (historical) Core Image path and
        // the Android module.
        AsyncFunction("applyBeauty") { (uri: String, _: Double) -> String in
            uri
        }

        View(BeautyCameraView.self) {
            // ~1/s with {fps}: the preview frame rate, for the debug indicator.
            Events("onFps")

            Prop("facing") { (view: BeautyCameraView, facing: String?) in
                view.setFacing(facing ?? "front")
            }

            Prop("smoothing") { (view: BeautyCameraView, value: Double?) in
                view.setSmoothing(Float(value ?? 0))
            }

            Prop("glow") { (view: BeautyCameraView, value: Double?) in
                view.setGlow(Float(value ?? 0))
            }

            Prop("clarity") { (view: BeautyCameraView, value: Double?) in
                view.setClarity(Float(value ?? 0))
            }

            Prop("warmth") { (view: BeautyCameraView, value: Double?) in
                view.setWarmth(Float(value ?? 0))
            }

            Prop("sharpness") { (view: BeautyCameraView, value: Double?) in
                view.setSharpness(Float(value ?? 0))
            }

            Prop("eyeEnlarge") { (view: BeautyCameraView, value: Double?) in
                view.setEyeEnlarge(Float(value ?? 0))
            }

            Prop("noseSlim") { (view: BeautyCameraView, value: Double?) in
                view.setNoseSlim(Float(value ?? 0))
            }

            Prop("faceSlim") { (view: BeautyCameraView, value: Double?) in
                view.setFaceSlim(Float(value ?? 0))
            }

            Prop("mustache") { (view: BeautyCameraView, enabled: Bool?) in
                view.setMustache(enabled == true)
            }

            Prop("faceMesh") { (view: BeautyCameraView, enabled: Bool?) in
                view.setFaceMesh(enabled == true)
            }

            AsyncFunction("takePicture") { (view: BeautyCameraView, promise: Promise) in
                view.takePicture(promise)
            }
        }
    }
}

File: modules/beauty-filter/android/src/main/java/com/haywan/filtercam/beautyfilter/BeautyFilterModule.kt
package com.haywan.filtercam.beautyfilter

import com.haywan.filtercam.beautyfilter.view.BeautyCameraView
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Expo module entry point. Declares the JS-facing contract:
 *  - `isAvailable` / `applyBeauty` module functions, and
 *  - the `BeautyFilter` view with its props and `takePicture` method.
 *
 * All rendering, camera and tracking work lives under the `view`, `render`,
 * `tracking` and `gl` sub-packages; this file is only the wiring.
 */
class BeautyFilterModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("BeautyFilter")

        Function("isAvailable") { true }

        // On Android the beauty filter is applied live in the GL pipeline and
        // captures already include it, so post-processing is a no-op. This
        // exists for API parity with the iOS module.
        AsyncFunction("applyBeauty") { uri: String, _: Double ->
            uri
        }

        View(BeautyCameraView::class) {
            // ~1/s with {fps}: the preview frame rate, for the debug indicator.
            Events("onFps")

            Prop("facing") { view, facing: String? ->
                view.setFacing(facing ?: "front")
            }

            Prop("smoothing") { view, value: Float? ->
                view.setSmoothing(value ?: 0f)
            }

            Prop("glow") { view, value: Float? ->
                view.setGlow(value ?: 0f)
            }

            Prop("clarity") { view, value: Float? ->
                view.setClarity(value ?: 0f)
            }

            Prop("warmth") { view, value: Float? ->
                view.setWarmth(value ?: 0f)
            }

            Prop("sharpness") { view, value: Float? ->
                view.setSharpness(value ?: 0f)
            }

            Prop("eyeEnlarge") { view, value: Float? ->
                view.setEyeEnlarge(value ?: 0f)
            }

            Prop("noseSlim") { view, value: Float? ->
                view.setNoseSlim(value ?: 0f)
            }

            Prop("faceSlim") { view, value: Float? ->
                view.setFaceSlim(value ?: 0f)
            }

            Prop("mustache") { view, enabled: Boolean? ->
                view.setMustache(enabled == true)
            }

            Prop("faceMesh") { view, enabled: Boolean? ->
                view.setFaceMesh(enabled == true)
            }

            AsyncFunction("takePicture") { view: BeautyCameraView, promise: Promise ->
                view.takePicture(promise)
            }
        }
    }
}
